[TRTLLM-3442] feat: added beam search support to the PyTorch Workflow

stnie opened the pull request:
## Description

- Added Basic Beam Search support for the PyTorch Workflow. 
- Return Context Logits and Generation Logits support for beam search is provided
- Return Logprobs Support is only provided for a single prompt
- Tests were run using TinyLlama model 

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
feat: added beam search support to the PyTorch Workflow

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
test: add unit tests for beam search functionality in the PyTorch workflow

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
fix: update decoder state access in TRTLLMSampler for beam search finalization

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

dcampora commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9403](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9403/) [ run ] triggered by Bot

dcampora reviewed (changes requested):
Looks good to me, please do the changes required and let's merge it.

tensorrt-cicd commented:
[PR_Github #9403](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9403/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6899](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6899/) completed with status: 'FAILURE'

stnie committed:
chore: add cache indirection properties to DecoderState and update TRTLLMSampler for improved beam search handling

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -333,7 +334,7 @@ def add_dummy_requests(
         max_num_draft_tokens: int = 0,
     ):
-        beam_width = 1
+        beam_width = 1  # TODO: more than 1 beam?
```
During warmup, I think it can stay at 1.

stnie committed:
chore: remove unnecessary non_blocking flags in tensor operations for cleaner code

Updated tensor operations in model_engine.py, resource_manager.py, and sampler.py to eliminate redundant non_blocking flags. Removed unnecessary parameter in unittest.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on tensorrt_llm/_torch/pyexecutor/sampler.py:
```diff
@@ -762,3 +783,81 @@ def update_requests(self, state: SampleStateTRTLLM):
+            log_probs_host = self.store["decoder_state"].log_probs.to(
+                'cpu', non_blocking=False)
+            cum_log_probs_host = self.store["decoder_state"].cum_log_probs.to(
+                'cpu', non_blocking=False)
```
done (here and in other cases)

stnie commented on tests/unittest/_torch/test_beam_search.py:
```diff
@@ -0,0 +1,96 @@
+                                   max_tokens: int, num_prompts: int):
+    if not enable_trtllm_sampler:
+        pytest.skip(
+            "Beam search currently is only supported with TRTLLMSampler")
```
done

tensorrt-cicd commented:
[PR_Github #9722](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9722/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7157](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7157/) completed with status: 'FAILURE'

stnie committed:
fix: update beam search unit test to use fixed TRTLLM sampler parameter

Removed the parameter for enabling TRTLLM sampler in the test function and set it to True for consistency in output shape validation.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

Funatiq reviewed (changes requested):
Do we have a check somewhere for `executor_config.max_beam_width == req_beam_width`? What happens if this is not the case? It would be good to have a test for that.

stnie committed:
chore: streamline beam search output handling in quickstart_advanced.py and improve tensor operations in various modules

Updated the output handling in quickstart_advanced.py to simplify the printing of generated text and associated logits based on beam search. Refactored tensor operations in model_engine.py, resource_manager.py, and sampler.py for clarity and efficiency. Added assertions to ensure beam width consistency in request handling.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
fix: ensure correct request handling in TRTLLMSampler by using scheduled_requests directly

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
fix: corrected handling of previous batch indices in PyTorchModelEngine for beam search

Updated the logic to ensure previous batch indices are  appended for each beam during request processing

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on tensorrt_llm/_torch/attention_backend/trtllm.py:
```diff
@@ -482,6 +482,12 @@ def is_nvfp4_output_kernel_available(
 
+    # TrtllmAttention needs to know the beam width and access to the cache indirection buffer,
+    # when beam search is enabled.
+    # TODO -- put this into the parent class instead?
```
I left in in the TrtllmAttentionMetadata and removed the todo

stnie commented on tensorrt_llm/_torch/pyexecutor/llm_request.py:
```diff
@@ -138,6 +138,16 @@ def append(self,
+        log_probs: [beam_width, num_tokens]
+        cum_log_probs: [beam_width]
+        """
+        self.beam_width = -1
```
I previously relied on the fact, that append, reinitializes the buffer if beam_width is -1. However, for clarity I directly call the init function instead of setting beam_width to -1.

tensorrt-cicd commented:
[PR_Github #10029](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10029/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #7400](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7400/) completed with status: 'FAILURE'

stnie committed:
chore: integrate max_beam_width parameter into executor and resource management

Added max_beam_width assertion in the autodeploy executor and added max_beam_width in the PyExecutor class. Updated related assertions and resource management to ensure consistent handling of beam width across components.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
chore: Update ADEngine and readability in ModelEngine

Added cache_indirection_buffer parameter to ADEngine forward method and adjusted integration of max_beam_width in create_autodeploy_executor to support enhanced functionality in the execution flow.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #10142](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10142/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #10142](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10142/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7488](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7488/) completed with status: 'SUCCESS'

stnie committed:
chore: updated format

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #10164](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10164/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7503](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7503/) completed with status: 'FAILURE'

dcampora commented on tests/integration/test_lists/test-db/l0_a30.yml:
```diff
@@ -18,6 +18,7 @@ l0_a30:
   - unittest/_torch/modeling -k "modeling_qwen"
   - unittest/_torch/modeling -k "modeling_qwen_moe"
   - unittest/_torch/auto_deploy/unit/singlegpu
+  - unittest/_torch/test_beam_search.py
```
No need to add it here.

Funatiq replied:
IIRC @stnie mentioned that it doesn't run if it isn't added. Have we confirmed that it runs in another stage?

stnie committed:
chore: enforce max_beam_width constraint in autodeploy executor and correct decoder input buffer assignment in sampler

Added an assertion to ensure max_beam_width is limited to 1 in the create_autodeploy_executor function. Updated the naming of decoder input buffers in TRTLLMSampler to use the correctg buffer.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10347](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10347/) [ run ] completed with state `ABORTED`

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10348](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10348/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `ab0fda3`

stnie committed:
chore: explicitly add test case for beam search functionality in l0_a30 integration tests

Included a new test entry for unittest/_torch/test_beam_search.py in the l0_a30.yml configuration to enable testing of beam search features.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10377](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10377/) [ run ] triggered by Bot

stnie committed:
refactor: update PyTorchModelEngine and TRTLLMSampler for improved beam search handling

Removed the static BEAM_WIDTH attribute from PyTorchModelEngine and adjusted the handling of previous batch indices to ensure they are appended correctly per request in beam search mode.  Additionally, modified the test for beam search output shapes to accommodate differing output and processing beam counts.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

Funatiq commented:
/bot run --disable-fail-fast

stnie committed:
chore: enhance attention metadata and resource management for beam search

Added max_num_sequences to AttentionMetadata and updated related classes to utilize this new parameter. Adjusted the PyTorchModelEngine to calculate max_num_sequences based on batch size and max_beam_width. Updated KVCacheManager and other components to ensure consistent handling of beam during kv cache size estimation. Added an exception when running beam search with overlap scheduler and adjusted test_beam_search accordingly.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10652](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10652/) [ run ] triggered by Bot

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10779](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10779/) [ run ] triggered by Bot

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10823](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10823/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #10828](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10828/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #10828](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10828/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #7999](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7999/) completed with status: 'FAILURE'

stnie committed:
fix: set default value for max_num_sequences in AttentionMetadata and make it optional

Updated the AttentionMetadata class to set a default value for max_num_sequences to None. Adjusted the TrtllmAttentionMetadata class to assign max_num_sequences based on max_num_requests if it is not explicitly set. This change ensures that beam search with TrtllmAttention Backend runs, without affecting single Beam cases or other backends

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10839](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10839/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #10839](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10839/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #8008](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8008/) completed with status: 'FAILURE'

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10958](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10958/) [ run ] triggered by Bot

stnie committed:
chore: updated code to be in line with coding style guidelines

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented:
/bot run

tensorrt-cicd commented:
[PR_Github #10969](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10969/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8103](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8103/) completed with status: 'FAILURE'

stnie commented:
/bot run --disable-fail-fast

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11011](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11011/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11011](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11011/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8136](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8136/) completed with status: 'SUCCESS'
